File: bits/bits_ans.pl
#!perl

use strict;
use warnings;

## テーブルを参照して解答する
my $txt = "10_0-10_4.txt";
open my $th, '<', $txt or die "cannot open $txt... : $!";

my %table;
while( my $line = <$th> ) {
    chomp($line);
    my ($num, $cnt) = split(/ /, $line);
    $table{$num} = $cnt;
}

## 問題ファイル
my $f = $ARGV[0];
if( !defined($f) ) { die "please specify the input file... : $!"; }
open my $fh, '<', $f or die "cannot open $f... : $!";

## 問題数
my $T = <$fh>; #print $T;
chomp($T);

## 解答
my $case = 1;
while( $case <= $T ) {

    my $N = <$fh>;
    my $P = f($N);

    print "Case #$case: $P\n";

    $case++;
}

sub f {
    my $num = shift;

    my $max = 0;

    for(my $i=0; $i <= int($num / 2); $i++ ) {
        my $cand = $table{$i} + $table{$num-$i};

        if( $max < $cand ) { $max = $cand; }
    }

    return $max;
}

File: bits/bits_ans.rb
#! /usr/bin/ruby
# -*- coding: utf-8 -*-

## bits count
def bits(num, table) 
  max = 0

  i_max = num / 2
  for i in 0..i_max
    cand = table[i] + table[num-i]
    if max < cand
      max = cand
    end
  end

  return max
end

## input
txt = ARGV[0]
th = open(txt, "r")

table = Hash.new
while line = th.gets
  num, cnt = split(" ")
  table[num.to_i] = cnt.to_i
end

## 問題ファイル
f = ARGV[1]
fh = open(f, "r")

## 問題数
t = fh.gets.chomp

## 解答
cnt = 1
while cnt <= t.to_i
  n = fh.gets.chomp.to_i
  c = bits(n, table)

  print 'Case #', cnt, ': ', c, "\n"
  
  cnt = cnt + 1
end
